services/api/service.go

```diff
@@ -835,6 +960,54 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 		if err != nil {
 			log.WithError(err).Error("failed to increment builder-stats after getPayload")
 		}
+
+		// Check if the block was valid.
+		submitBlockReq := types.BuilderSubmitBlockRequest{
+			Signature:        payload.Signature,
+			Message:          &bidTrace.BidTrace,
+			ExecutionPayload: getPayloadResp.Data,
+		}
+
+		simErr := api.simulateBlock(blockSimOptions{
+			ctx:      req.Context(),
+			log:      log,
+			highPrio: true, // manually set to true for these blocks.
+			req: &BuilderBlockValidationRequest{
+				BuilderSubmitBlockRequest: submitBlockReq,
+			},
+		})
+		// Block validation failed, perform a demotion.
+		if simErr != nil {
+			builderPubkey := bidTrace.BuilderPubkey.String()
+			log = log.WithFields(logrus.Fields{
+				"builderPubkey": builderPubkey,
+				"bidTrace":      bidTrace,
+			})
+			log.WithError(simErr).Error("simulation error")
+
+			// Prepare demotion data.
+			signedBeaconBlock := SignedBlindedBeaconBlockToBeaconBlock(payload, getPayloadResp.Data)
+
+			// Get registration entry from the DB.
+			registrationEntry, err := api.db.GetValidatorRegistration(builderPubkey)
+			if err != nil {
+				if errors.Is(err, sql.ErrNoRows) {
+					log.WithError(err).Error("no registration found for validator " + builderPubkey)
+				} else {
+					log.WithError(err).Error("error reading validator registration")
+				}
+			}
+			var signedRegistration *types.SignedValidatorRegistration
+			if registrationEntry != nil {
+				signedRegistration, err = registrationEntry.ToSignedValidatorRegistration()
+				if err != nil {
+					log.WithError(err).Error("error converting registration to signed registration")
+				}
+			}
+
+			api.demoteBuilder(builderPubkey, &submitBlockReq, signedBeaconBlock, signedRegistration)
```

Comment:
Adding a demotion entry for non-optimistic blocks may lead to a bunch of false-positive demotion entries.